src/Components/Server/src/Circuits/RemoteNavigationManager.cs

```diff
@@ -89,21 +89,28 @@ protected override void NavigateToCore(string uri, NavigationOptions options)
             throw new NavigationException(absoluteUriString);
         }
 
-        NotifyLocationChangingAsync(uri, options.HistoryEntryState, false).AsTask().ContinueWith(t =>
+        _ = PerformNavigationAsync();
+
+        async Task PerformNavigationAsync()
         {
-            if (t.Exception is { } exception)
-            {
-                Log.NavigationFailed(_logger, uri, exception);
-            }
-            else if (!t.Result)
+            try
             {
-                Log.NavigationCanceled(_logger, uri);
+                var shouldContinueNavigation = await NotifyLocationChangingAsync(uri, options.HistoryEntryState, false);
+
+                if (shouldContinueNavigation)
+                {
+                    await _jsRuntime.InvokeVoidAsync(Interop.NavigateTo, uri, options);
+                }
+                else
+                {
+                    Log.NavigationCanceled(_logger, uri);
+                }
             }
-            else
+            catch (Exception ex)
             {
-                _jsRuntime.InvokeVoidAsync(Interop.NavigateTo, uri, options).Preserve();
+                Log.NavigationFailed(_logger, uri, ex);
```

Comment:
What happens when a handler throws in this scenario compared to when the navigation is started from JS?

Reply:
At the time you wrote this comment, it would log an error but not tear down the circuit. I've since changed that behavior so that it tears down the circuit for programmatically initiated navigations like it does for JS-initiated navigations.